### PR TITLE
fix: flaky test

### PR DESCRIPTION
**Pull request by CaptainStandby**
This fixes a problematic unit test, that monitors log messages and expected a specific message at the end of the list. This has been corrected by searching for the expected message in the list.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: driver/config/config_test.go [hunk 1]
```diff
@@ -941,32 +940,31 @@ func TestIdentitySchemaValidation(t *testing.T) {
...
-		c := make(chan bool, 1)
+	testWatch := func(t *testing.T, ctx context.Context, cmd *cobra.Command, identity *configFile) (*config.Config, *test.Hook, func([]map[string]string)) {
 		tdir := t.TempDir()
 		assert.NoError(t,
 			os.MkdirAll(tdir, // DO NOT CHANGE THIS: https://github.com/fsnotify/fsnotify/issues/340
 				os.ModePerm))
 		configFileName := randx.MustString(8, randx.Alpha)
 		tmpConfig, err := os.Create(filepath.Join(tdir, configFileName+".config.yaml"))
 		assert.NoError(t, err)
+		t.Cleanup(func() { tmpConfig.Close() })
```
Comment (alnr) [inline driver/config/config_test.go:951]: nit
```suggestion
		t.Cleanup(tmpConfig.Close)
```

**Review by alnr — APPROVED** [review]

File: driver/config/config_test.go (hunk 1 above)
Comment (alnr) [reply driver/config/config_test.go]: I'm an idiot

**Review by alnr — APPROVED** [review]

File: driver/config/config_test.go (hunk 1 above)
Comment (aeneasr) [reply driver/config/config_test.go:951]: The suggestion is correct :)

File: driver/config/config_test.go [hunk 2]
```diff
@@ -1022,34 +1020,44 @@ func TestIdentitySchemaValidation(t *testing.T) {
...
-					assert.NoError(t, err)
-
-					assert.Contains(t, lastHook, "The changed identity schema configuration is invalid and could not be loaded.")
+					// Change the identity config to an invalid file
+					writeSchema(invalidIdentity.Identity.Schemas)
+				}()
+
+				// There are a bunch of log messages beeing logged. We are looking for a specific one.
+				timeout := time.After(time.Millisecond * 500)
+			outer:
```
Comment (aeneasr) [inline driver/config/config_test.go:1040]: If possible try to avoid labels for better readability :)

File: driver/config/config_test.go [hunk 3]
```diff
@@ -1022,34 +1020,44 @@ func TestIdentitySchemaValidation(t *testing.T) {
...
+						}
+
+						if strings.Contains(s, "The changed identity schema configuration is invalid and could not be loaded.") {
+							break outer
+						}
+					}
+
+					select {
+					case <-ctx.Done():
+						panic("the test could not complete as the context timed out before the file watcher updated")
```
Comment (aeneasr) [inline driver/config/config_test.go:1056]: No panics ever please

File: driver/config/config_test.go [hunk 4]
```diff
@@ -1022,34 +1020,44 @@ func TestIdentitySchemaValidation(t *testing.T) {
...
+				}()
+
+				// There are a bunch of log messages beeing logged. We are looking for a specific one.
+				timeout := time.After(time.Millisecond * 500)
+			outer:
+				for {
+					for _, v := range hook.AllEntries() {
+						s, err := v.String()
+						if err != nil {
+							t.Errorf("Unexpected Error: %s", err.Error())
```
Comment (aeneasr) [inline driver/config/config_test.go:1045]: Please use `require.NoError` or `assert.NoError`

**Comment by aeneasr** [comment]
@CaptainStandby can you please address the comments in a follow-up PR? Thank you!

**Comment by CaptainStandby** [comment]
Sure thing #2964

**Review by zepatrik — COMMENTED** [review]
I also investigated this already. The underlying issue is that `marshalAndWrite` causes two notifications (one for `.Truncate` and one for `.Write`).
The file content is then read after the first event. Depending on the speed of the system/disk this can then flake, returning invalid content (i.e. empty file), or the expected content. Ideally we would do an [atomic file write](https://github.com/kubernetes/kubernetes/blob/03bfbdd8aa04739a5d99bdefeb9fcf379ffb970d/pkg/volume/util/atomic_writer.go) in the test instead, but this is also OK.

**Comment by CaptainStandby** [comment]
Yeah, we noticed this too and the first take on fixing this, was doing atomic file write. But we decided to go with this approach, because we wanted to retain the "flakyness" of the write to test that the system could correctly deal with that situation.
Maybe a specific test, that simulates a problematic write could be added, now that I think about it.

**Comment by alnr** [comment]
Yes, the non-atomic write is almost certainly the problem.

Like @CaptainStandby said, we purposefully kept that test behavior because non-atomic writes could easily happen in real-world use, too. Instead of changing the test to use an atomic move, we more broadly inspect what's logged.

A similar flakyness could happen if the logger internally re-ordered log lines.

**Pull request merged**